ra-optimisation/search-tree-rao/src/main/java/com/powsybl/open_rao/search_tree_rao/castor/algorithm/AutomatonSimulator.java

```diff
@@ -572,43 +572,33 @@ RangeAutomatonSimulationResult shiftRangeActionsUntilFlowCnecsSecure(List<RangeA
         double sensitivityUnderestimator = 1;
         while (!flowCnecsWithNegativeMargin.isEmpty()) {
             FlowCnec toBeShiftedCnec = flowCnecsWithNegativeMargin.get(0).getLeft();
-            if (toBeShiftedCnec.equals(previouslyShiftedCnec)) {
-                sensitivityUnderestimator = Math.max(SENSI_UNDER_ESTIMATOR_MIN, sensitivityUnderestimator - SENSI_UNDER_ESTIMATOR_DECREMENT);
-            } else {
-                sensitivityUnderestimator = 1;
-            }
+
+            sensitivityUnderestimator = updateSensitivityUnderestimator(toBeShiftedCnec, previouslyShiftedCnec, sensitivityUnderestimator);
+
             Side side = flowCnecsWithNegativeMargin.get(0).getRight();
-            // Aligned range actions have the same set-point :
-            double currentSetpoint = alignedRangeActions.get(0).getCurrentSetpoint(network);
-            double optimalSetpoint = currentSetpoint;
-            double conversionToMegawatt = RaoUtil.getFlowUnitMultiplier(toBeShiftedCnec, side, flowUnit, MEGAWATT);
-            double cnecFlow = conversionToMegawatt * automatonRangeActionOptimizationSensitivityAnalysisOutput.getFlow(toBeShiftedCnec, side, flowUnit);
-            double cnecMargin = conversionToMegawatt * automatonRangeActionOptimizationSensitivityAnalysisOutput.getMargin(toBeShiftedCnec, side, flowUnit);
-            double sensitivityValue = 0;
-            // Under-estimate range action sensitivity if convergence to margin = 0 is slow (ie if multiple passes
-            // through this loop have been needed to secure the same CNEC)
-            for (RangeAction<?> rangeAction : alignedRangeActions) {
-                sensitivityValue += sensitivityUnderestimator * automatonRangeActionOptimizationSensitivityAnalysisOutput.getSensitivityValue(toBeShiftedCnec, side, rangeAction, MEGAWATT);
-            }
+            double sensitivityValue = computeTotalSensitivityValue(alignedRangeActions, sensitivityUnderestimator, automatonRangeActionOptimizationSensitivityAnalysisOutput, toBeShiftedCnec, side);
+
             // if sensitivity value is zero, CNEC cannot be secured. move on to the next CNEC with a negative margin
             if (Math.abs(sensitivityValue) < DOUBLE_NON_NULL) {
                 flowCnecsToBeExcluded.add(Pair.of(toBeShiftedCnec, side));
                 flowCnecsWithNegativeMargin = getCnecsWithNegativeMarginWithoutExcludedCnecs(flowCnecs, flowCnecsToBeExcluded, automatonRangeActionOptimizationSensitivityAnalysisOutput, activatedRangeActionsWithSetpoint);
                 continue;
             }
 
-            double optimalSetpointForSide = computeOptimalSetpoint(currentSetpoint, cnecFlow, cnecMargin, sensitivityValue, alignedRangeActions.get(0), minSetpoint, maxSetpoint);
-            if (Math.abs(currentSetpoint - optimalSetpointForSide) > Math.abs(currentSetpoint - optimalSetpoint)) {
-                optimalSetpoint = optimalSetpointForSide;
-            }
```

Comment:
where did this go?

Reply:
I believe this was some old code that was forgotten during a refactoring.

optimalSetpoint is set to currentSetpoint so the right side is equal to 0. 
So either currentSetpoint is different to optimalSetpointForSide and so the abs(diff) > 0 and we should set optimalSetpoint to optimalSetpointForSide
Or optimalSetpointForSide = currentSetpoint (= optimalSetpoint), and so setting optimalSetpoint to optimalSetpointForSide doesn't change anything.

(I guess at some point we used to check the optimal setpoint for each side, and we were applying the setpoint that required the bigger change to make sure each side is secure.)